### ISSUE TITLE
Git and repository caching issues with 2.1.0+?

### ISSUE THREAD

**Issue by booleanbetrayal** [argoproj/argo-cd#7230]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

We are seeing the following error occasionally, resulting in ArgoCD Sync failures that propagate as Unknown sync status for any Application that relies on that repo:

```
ComparisonError: rpc error: code = Internal desc = Failed to checkout b5e7834eb98b8887579036caf7d5db5cfa404772: `git lfs ls-files -n` failed exit status 128: Error: failed to call git rev-parse --git-dir --show-toplevel: ""
```

This happens sporadically and we believe it began happening in 2.1.0 (having not seen it in 2.0.0) and may be related to the following release announcements:

> - The synchronization process became much much faster and requires significantly less memory.
> - An additional caching that ensures that each repository's target revisions are queried only once per
reconciliation cycle. This dramatically reduces the number of Git requests.

When we encounter this situation, restarting the ArgoCD redis and reposerver pods resolves the issue.

**To Reproduce**

At this point, we don't have an upstream explanation. We generally push to our git-ops repo and immediately instruct ArgoCD to synchronize with those changes.

**Expected behavior**

Syncing of recent commits continues to work as expected without error.

**Screenshots**

<img width="1439" alt="Screen Shot 2021-09-15 at 11 02 44 AM" src="https://user-images.githubusercontent.com/981825/133477634-22e5d519-d671-4374-915b-c0bee8c10bb4.png">

**Version**

```shell
argocd: v2.1.2+7af9dfb
  BuildDate: 2021-09-02T18:05:23Z
  GitCommit: 7af9dfb3524c13e941ab604e36e49a617fe47d2e
  GitTreeState: clean
  GoVersion: go1.16.5
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

Here are some (sanitized) logs from our logging aggregation on the repo-server container
```
"2021-09-15T16:48:52.252305763Z stderr F time=\"2021-09-15T16:48:52Z\" level=info msg=Trace args=\"[git checkout --force b5e7834eb98b8887579036caf7d5db5cfa404772]\" dir=/tmp/git@some-internal-repository-alias operation_name=\"exec git\" time_ms=3.2188120000000002"
"2021-09-15T16:48:52.252646853Z stderr F time=\"2021-09-15T16:48:52Z\" level=info msg=\"git lfs ls-files -n\" dir=/tmp/git@some-internal-repository-alias execID=8hOfk"
"2021-09-15T16:48:52.267516146Z stderr F time=\"2021-09-15T16:48:52Z\" level=error msg=\"`git lfs ls-files -n` failed exit status 128: Error: failed to call git rev-parse --git-dir --show-toplevel: \\\"\\\"\" execID=8hOfk"
"2021-09-15T16:48:52.268191356Z stderr F time=\"2021-09-15T16:48:52Z\" level=info msg=Trace args=\"[git lfs ls-files -n]\" dir=/tmp/git@some-internal-repository-alias operation_name=\"exec git\" time_ms=15.070175"
"2021-09-15T16:48:52.270238255Z stderr F time=\"2021-09-15T16:48:52Z\" level=info msg=\"git fetch origin --tags --force\" dir=/tmp/git@some-internal-repository-alias execID=OQArP"
"2021-09-15T16:48:52.344057124Z stderr F time=\"2021-09-15T16:48:52Z\" level=error msg=\"finished unary call with code Internal\" error=\"rpc error: code = Internal desc = Failed to checkout b5e7834eb98b8887579036caf7d5db5cfa404772: `git lfs ls-files -n` failed exit status 128: Error: failed to call git rev-parse --git-dir --show-toplevel: \\\"\\\"\" grpc.code=Internal grpc.method=GenerateManifest grpc.request.deadline=\"2021-09-15T16:49:39Z\" grpc.service=repository.RepoServerService grpc.start_time=\"2021-09-15T16:48:39Z\" grpc.time_ms=12314.345 span.kind=server system=grpc"
"2021-09-15T16:48:52.353045013Z stderr F time=\"2021-09-15T16:48:52Z\" level=info msg=\"manifest cache miss: &ApplicationSource{RepoURL:git@repository-provider.com:/org/repository,Path:some-path,TargetRevision:main,Helm:&ApplicationSourceHelm{ValueFiles:[some-values.yaml],Parameters:[]HelmParameter{},ReleaseName:,Values:,FileParameters:[]HelmFileParameter{},Version:,},Kustomize:nil,Ksonnet:nil,Directory:nil,Plugin:nil,Chart:,}/b5e7834eb98b8887579036caf7d5db5cfa404772\""
"2021-09-15T16:48:53.550696511Z stderr F time=\"2021-09-15T16:48:53Z\" level=info msg=Trace args=\"[git fetch origin --tags --force]\" dir=/tmp/git@some-internal-repository-alias operation_name=\"exec git\" time_ms=1280.218772"
"2021-09-15T16:48:53.550725372Z stderr F time=\"2021-09-15T16:48:53Z\" level=info msg=\"git lfs ls-files -n\" dir=/tmp/git@some-internal-repository-alias execID=Mp7Ez"
"2021-09-15T16:48:53.560026036Z stderr F time=\"2021-09-15T16:48:53Z\" level=error msg=\"`git lfs ls-files -n` failed exit status 128: Error: failed to call git rev-parse --git-dir --show-toplevel: \\\"\\\"\" execID=Mp7Ez"
"2021-09-15T16:48:53.560055705Z stderr F time=\"2021-09-15T16:48:53Z\" level=info msg=Trace args=\"[git lfs ls-files -n]\" dir=/tmp/git@some-internal-repository-alias operation_name=\"exec git\" time_ms=9.357464"
"2021-09-15T16:48:53.560062367Z stderr F time=\"2021-09-15T16:48:53Z\" level=info msg=\"git checkout --force b5e7834eb98b8887579036caf7d5db5cfa404772\" dir=/tmp/git@some-internal-repository-alias execID=3IVlQ"
```

**Comment by alexmt** [comment]
Most likely it is related to:

> An additional caching that ensures that each repository's target revisions are queried only once per
reconciliation cycle. This dramatically reduces the number of Git requests.

Argo CD caches `git ls-remote` for 3 mins. @booleanbetrayal how do you think is it possible that the repository got `git push --force` and some commits got removed from git history?

**Comment by booleanbetrayal** [comment]
@alexmt - We have a squash-merge policy on our repo for feature branches going into `main` (which kicks off our git-ops flow). Likewise, developers are free to rebase at will in those feature branches. Both those processes would definitely be dropping commits, but no history rewriting should be happening in `main`. We actually have a repo policy that forbids any force push operations there.

**Comment by booleanbetrayal** [comment]
@alexmt - Just thinking out loud, but we sometimes spool ArgoCD environments for these feature branches ... I'm wondering if having an environment polling that git branch (set as the `--revision` values) at some point, while it is in the process of being squash merged into `main` could bring this about. 

The net result is that all the environments that are polling that repo, regardless of target revision, all seem to go into status of Unknown.  Maybe there needs to be some sort of cache kick and retry under this error scenario?

**Comment by alexmt** [comment]
I think you are right @booleanbetrayal . Trying to reproduce it locally and will prioritize fixing it.

**Comment by booleanbetrayal** [comment]
Thanks @alexmt !

**Comment by booleanbetrayal** [comment]
Sounds like #7244 may close this one out. Will update us shortly and keep an eye on it.

**Comment by alexmt** [comment]
I've spent a couple of days trying to reproduce it but no luck so far. Discovered that in the case of the manual sync the API server retrieves commit SHA without using the cache (API server doesn't even have access to revisions cache at all).

> resulting in ArgoCD Sync failures that propagate as Unknown sync status for any Application that relies on that repo
> When we encounter this situation, restarting the ArgoCD redis and repo server pods resolves the issue.

This made me think that revision caching is probably not the root cause. Sounds like something happened with cloned git repo in the repo-server pod and HEAD resolves to invalid commit. Investigating this


@booleanbetrayal ,  unless it is secret, can you share which Git provider you are using?

**Comment by booleanbetrayal** [comment]
Okay @alexmt - I'll keep my eyes peeled for any recurrence now that we are on 2.1.3.

FWIW - We are on GitHub

**Comment by alexmt** [comment]
Tried to research what else has changed in 2.1. We actually upgraded git and git-lfs versions:

* v2.0.0
```
$ git --version
git version 2.27.0
$ git lfs env
git-lfs/2.13.2 (GitHub; linux amd64; go 1.15.4)
```
* v2.1.3
```
$ git --version
git version 2.30.2
$ git lfs env
git-lfs/2.11.0 (GitHub; linux amd64; go 1.14.2)
```

It might be that error is related to git LFS only. @booleanbetrayal have you seen the Unknown errors for apps that are based on repos without LFS ?

**Comment by booleanbetrayal** [comment]
We don't actually have any apps built around repos without LFS, unfortunately, @alexmt . I think only LFS objects for us would be Helm chart binaries.

**Comment by alexmt** [comment]
Thank you, @booleanbetrayal!

According to logs, the error happens only if LFS is enabled. Since you don't need LFS can you try to disable it for the repository to confirm it is only LFS related? The LFS can be disabled in repo settings by adding `enableLfs: "false"` key to repo secret data.

**Comment by booleanbetrayal** [comment]
Will give that a whirl! I had thought we were using LFS in our gitops repo, but it's actually not reflecting any stored files, so it was never initialized as such. Secrets are updated and I'll keep my eyes on it!

**Comment by booleanbetrayal** [comment]
Just a note that we just re-experienced this after updating to `v2.1.6` and re-enabling LFS support. Going to keep LFS support off for the timing being.
